### PR TITLE
Rewrite TensorImpl to use TensorTypeId.

### PR DESCRIPTION
**Pull request by ezyang**
Summary:
Translation to Backend happens immediately before we go into the
Type universe; otherwise we use TensorTypeId.

I allocated TensorTypeId corresponding exactly to existing ATen
Backend.  Only CPUTensorId and CUDATensorId are relevant in the
Caffe2 universe.

Differential Revision: D9184060

### CODE REVIEW

File: aten/src/ATen/Formatting.cpp [hunk 1]
```diff
@@ -45,10 +45,6 @@ std::ostream& operator<<(std::ostream & out, Backend b) {
   return out << toString(b);
 }
 
-std::ostream& operator<<(std::ostream & out, ScalarType t) {
```
Comment (gchanan) [inline aten/src/ATen/Formatting.cpp]: why this change?

File: aten/src/ATen/SparseTensorImpl.cpp [hunk 2]
```diff
@@ -18,15 +30,13 @@ namespace at {
...
+SparseTensorImpl::SparseTensorImpl(at::TensorTypeId type_id, at::ScalarType scalar_type)
+    : TensorImpl(type_id, scalar_type, nullptr, false)
     , size_{0}
     , sparseDims_(1)
     , denseDims_(0)
-    , indices_(globalContext().getTypeOpt(toDense(backend), ScalarType::Long)->tensor())
-    , values_(globalContext().getTypeOpt(toDense(backend), scalar_type)->tensor()) {
-      AT_ASSERT(backend == Backend::SparseCPU || backend == Backend::SparseCUDA);
-    }
+    , indices_(globalContext().getTypeOpt(sparseTensorIdToDenseBackend(type_id), ScalarType::Long)->tensor())
```
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.cpp]: this will conflict with https://github.com/pytorch/pytorch/pull/9279 -- can you work out ordering with @yf225?

File: aten/src/ATen/core/TensorTypeId.cpp [hunk 3]
```diff
@@ -1,5 +1,9 @@
 #include "ATen/core/TensorTypeId.h"
 
+namespace at {
```
Comment (gchanan) [inline aten/src/ATen/core/TensorTypeId.cpp]: related to the other comment, why don't we do the opposite and move this to Formatting?  Is it because there are external usages that break?

**Review by gchanan — APPROVED** [review]

File: aten/src/ATen/Formatting.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/Formatting.cpp]: The short answer is, good C++ practice is to put `operator<<` for user defined types in the same namespace as the type in question.

The long answer is, C++ has a feature called argument dependent lookup, whereby when it is looking for an implementation of `operator<<` to satisfy something like `std::cout << scalar_type`, it collects the namespaces of all of its arguments and looks for `operator<<` in those namespaces. This is why this program compiles: https://godbolt.org/g/a4ZQyK even though, ordinarily, `f::operator<<` would not be considered "in-scope" for a normal function call.

Now, in many cases, whether or not you put the `operator<<` definition in the global namespace or in your user namespace doesn't matter, because scope resolution will find the right one anyway. But if you have some `operator<<` definitions in a namespace, and you perform a `<<` in that namespace, the compiler will see that the namespaced `operator<<` is the relevant one, and will ignore your overloads in the global namespace. Example: https://godbolt.org/g/eZjXEs

You can "work around" this problem by explicitly telling the compiler to use the global `operator<<` with a using declaration https://godbolt.org/g/PjvoSR but there is a simpler, more excellent way to fix the problem: DON'T define `operator<<` in the global namespace; instead, rely on ADL to figure out what the correct namespace of `operator<<` should be.

(Also CC @cpuhrsch, this is why you should do it this way!)

File: aten/src/ATen/core/TensorTypeId.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/core/TensorTypeId.cpp]: I think it's silly to have to include `Formatting.h` if you want to `<<` types; the implementations are really simple, put them with the actual headers.

File: aten/src/ATen/SparseTensorImpl.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/SparseTensorImpl.cpp]: I'll do it after that PR lands.
